src/zarr/core/metadata/converter/migrate_to_v3.py

```diff
@@ -0,0 +1,303 @@
+        if force or await _metadata_exists(
+            cast(Literal[2, 3], alternative_metadata), store_path / parent_path
+        ):
+            logger.info("Deleting metadata at %s", store_path / file_path)
+            if not dry_run:
+                awaitables.append((store_path / file_path).delete())
+        else:
+            raise ValueError(
+                f"Cannot remove v{zarr_format} metadata at {store_path / file_path} - no v{alternative_metadata} "
+                "metadata exists. To delete anyway, use the 'force' option."
+            )
+
+    await asyncio.gather(*awaitables)
+
+
+def _convert_group(zarr_v2: Group, output_path: StorePath, dry_run: bool) -> None:
+    # process members of the group
+    for key in zarr_v2:
+        migrate_to_v3(zarr_v2[key], output_path=output_path / key, dry_run=dry_run)
+
+    # write group's converted metadata
+    group_metadata_v3 = GroupMetadata(
+        attributes=zarr_v2.metadata.attributes, zarr_format=3, consolidated_metadata=None
+    )
+    sync(_save_v3_metadata(group_metadata_v3, output_path, dry_run=dry_run))
+
+
+def _convert_array(zarr_v2: Array, output_path: StorePath, dry_run: bool) -> None:
+    array_metadata_v3 = _convert_array_metadata(cast(ArrayV2Metadata, zarr_v2.metadata))
+    sync(_save_v3_metadata(array_metadata_v3, output_path, dry_run=dry_run))
+
+
+async def _metadata_exists(zarr_format: ZarrFormat, store_path: StorePath) -> bool:
+    metadata_files_required = {2: [ZARRAY_JSON, ZGROUP_JSON], 3: [ZARR_JSON]}
+
+    for metadata_file in metadata_files_required[zarr_format]:
+        if await (store_path / metadata_file).exists():
+            return True
+
+    return False
+
+
+def _convert_array_metadata(metadata_v2: ArrayV2Metadata) -> ArrayV3Metadata:
+    chunk_key_encoding = V2ChunkKeyEncoding(separator=metadata_v2.dimension_separator)
+
+    codecs: list[Codec] = []
+
+    # array-array codecs
+    if metadata_v2.order == "F":
+        # F is equivalent to order: n-1, ... 1, 0
+        codecs.append(TransposeCodec(order=list(range(len(metadata_v2.shape) - 1, -1, -1))))
+    codecs.extend(_convert_filters(metadata_v2))
+
+    # array-bytes codecs
+    if not isinstance(metadata_v2.dtype, HasEndianness):
+        codecs.append(BytesCodec(endian=None))
+    else:
+        codecs.append(BytesCodec(endian=metadata_v2.dtype.endianness))
+
+    # bytes-bytes codecs
+    bytes_bytes_codec = _convert_compressor(metadata_v2)
+    if bytes_bytes_codec is not None:
+        codecs.append(bytes_bytes_codec)
+
+    return ArrayV3Metadata(
+        shape=metadata_v2.shape,
+        data_type=metadata_v2.dtype,
+        chunk_grid=metadata_v2.chunk_grid,
+        chunk_key_encoding=chunk_key_encoding,
+        fill_value=metadata_v2.fill_value,
+        codecs=codecs,
+        attributes=metadata_v2.attributes,
+        dimension_names=None,
+        storage_transformers=None,
+    )
+
+
+def _convert_filters(metadata_v2: ArrayV2Metadata) -> list[ArrayArrayCodec]:
```

Comment:
I think it would make slightly more sense to pass filters, instead of the whole metadata here.
```suggestion
def _convert_filters(filters) -> list[ArrayArrayCodec]:
```
(I probably got the typing wrong)